gr-digital/lib/diversity_combiner_cc_impl.cc

```diff
@@ -0,0 +1,181 @@
+    {
+      return gnuradio::get_initial_sptr
+        (new diversity_combiner_cc_impl(num_inputs, vlen, combining_technique));
+    }
+
+    /*
+     * The private constructor
+     */
+    diversity_combiner_cc_impl::diversity_combiner_cc_impl(uint16_t num_inputs,
+                                                           uint16_t vlen,
+                                                           std::string combining_technique)
+      : gr::sync_block("diversity_combiner_cc",
+              gr::io_signature::make(num_inputs, num_inputs, vlen*sizeof(gr_complex)),
+              gr::io_signature::make(1, 1, vlen*sizeof(gr_complex))),
+        d_num_inputs(num_inputs),
+        d_vlen(vlen),
+        d_combining_technique(combining_technique),
+        d_best_path(0) // Initially, the SC algorithm selects channel 0, until it is changed by CSI.
+    {
+      // Resize vectors csi, csi_squared and mrc_weighting.
+      d_csi.resize(num_inputs);
+      d_csi_squared.resize(num_inputs);
+      d_mrc_weighting.resize(num_inputs);
+      // Initially, the MRC algorithm weights each channel equally, until it is changed by CSI.
+      std::fill(d_mrc_weighting.begin(), d_mrc_weighting.end(), 1.0/num_inputs);
+      // Set tag propagation policy to 'All to All'.
+      set_tag_propagation_policy(TPP_ALL_TO_ALL);
+    }
+
+    /*
+     * Our virtual destructor.
+     */
+    diversity_combiner_cc_impl::~diversity_combiner_cc_impl()
+    {
+    }
+
+    void
+    diversity_combiner_cc_impl::combine_inputs(gr_vector_const_void_star input,
+                                               gr_complex* out,
+                                               uint16_t offset,
+                                               uint16_t length){
+      // Calculate the magnitude square of the complex CSI vector.
+      for (int i = 0; i < d_num_inputs; ++i) {
+        d_csi_squared[i] = std::norm(d_csi[i]);
+      }
+      if(d_combining_technique.compare("SC") == 0) { // Selection combining
+        // Search for path coefficient with maximum magnitude.
+        d_best_path = std::distance(d_csi_squared.begin(),
+                                    std::max_element(d_csi_squared.begin(),
+                                                     d_csi_squared.end()));
+      }
+      else if(d_combining_technique.compare("MRC") == 0) { // Maximum-Ratio combining
+        // Calculate the normalized weighting coefficients.
+        float total_path_energy = std::accumulate(d_csi_squared.begin(), d_csi_squared.end(), 0.0);
+        for (int i = 0; i < d_num_inputs; ++i) {
+          d_mrc_weighting[i] = std::polar(std::sqrt(d_csi_squared[i]/total_path_energy), -std::arg(d_csi[i]));
+        }
+      }
+
+    }
+
+    void
+    diversity_combiner_cc_impl::process_symbol(gr_vector_const_void_star input,
+                                               gr_complex* out,
+                                               uint16_t offset,
```

Comment:
Why are offset and length just uint16_t? 64k is not small, but still in the range an application could get to. Better make this larger.